### ISSUE TITLE
User experience

### ISSUE THREAD

**Issue by NoSloppy** [ryryog25/ProffieConfig#13]
Still want notes?
Just grabbed 1.6.8. Looks good with the width issue fixed.
First I noticed (super nitpicky I know) that the text says "This is ProffieConfig's Main Menu". Maybe it should read. "To the right is ProffieConfig's Main Menu" instead of "This is.." ? . Just feels more intuitive being specific, as "This" refers to the window the user is reading the text in (the tutorial window).

Next, maybe add "seconds" after the PLI, Idle, and Motion timeout values to clarify what the time is in terms of (my brain thought in milliseconds initially)

Got up to the prop part. Button control list is nice, you've done a lot of work filtering by button etc. 
I suggest 2 buttons on this page: 
1. Print controls list or save as txt file for user's easy reference later (unless that's a thing I didn't get up to yet).
2. A button to open a readme section that is the descriptive (not button controls) section from the prop. For example, my prop has explanations of features that probably won't be understood without reading about it (like Monophonic Force or Swap for example.)
This is going to practically be required reading later as I added dual blade support.

I like that the style Editor included is the WIP version I've been tweaking (and that Dark Mode is defaulted to on (or maybe that's my cookies?). Anyway, I wish that could be finished up. I have stale submitted PRs waiting for review for a long time now. Almost there.

Now, problems:
Running through the end of the tutorial, the config didn't verify successfully:
![Screenshot 2024-05-16 at 12 43 00 AM](https://github.com/ryryog25/ProffieConfig/assets/53964195/17910970-1f4f-4b4c-b3ac-18ea3e74d652)
Here it is:( changed .h to .txt to attach here)
[balls3.txt](https://github.com/ryryog25/ProffieConfig/files/15329757/balls3.txt)

So I clicked OK, which opened the Main Menu window. I chose the config from the dropdown, then the app quit with an error popup "ProffieConfig quit unexpectedly."
Here's the crash report:
[crash.txt](https://github.com/ryryog25/ProffieConfig/files/15329775/crash.txt)

**Comment by NoSloppy** [comment]
Edit: I found the "Info" button with the descriptive prop features. Maybe "buttons..." should say "Button controls..." and then "Info..." button would be nested on that page instead? 
I don't know. That's where I feel I'd expect it maybe.

**Comment by NoSloppy** [comment]
Blade Awareness might need some work. I don't think you want to allow both detect and ID to be selected.

Enabling just Blade Detect, I can set Blade Detect Pin to something (17), but then can't edit anything else, where I would need to go to the Blade Arrays part and edit a blade_in and blade_out.
blade_out should use NO_BLADE.
Enabling Blade ID allows editing of that column, but blade_in doesn't allow editing of the name or ID resistor value (it's stuck at 0)
Yet, adding a blade_out array, both can be edited. Again, in the case where the array is to be used when no blade is inserted, NO_BLADE should be the value varable as per how ProffieOS uses it. (changes lately too... there's a range and stuff).
ALso, since Enable Power for Blade ID is required, maybe that should default to checkbox ON and greyed out, then allow the pin selection.

**Comment by ryryog25** [comment]
> First I noticed (super nitpicky I know) that the text says "This is ProffieConfig's Main Menu". Maybe it should read. "To the right is ProffieConfig's Main Menu" instead of "This is.." ? . Just feels more intuitive being specific, as "This" refers to the window the user is reading the text in (the tutorial window).

Makes sense to me, I've changed it.

> Next, maybe add "seconds" after the PLI, Idle, and Motion timeout values to clarify what the time is in terms of (my brain thought in milliseconds initially)

These are actually in minutes. I've added labels for them. (Any objections to them being in minutes?)

> Print controls list or save as txt file for user's easy reference later (unless that's a thing I didn't get up to yet)

Noted. wxWidgets has some tools for formatting things to be printed (and actually printing), I'll have to look into that. Would want it to look pretty after all :)

Saving it can also be an option I can add.

> A button to open a readme section that is the descriptive (not button controls) section from the prop. For example, my prop has explanations of features that probably won't be understood without reading about it (like Monophonic Force or Swap for example.)
I found the "Info" button with the descriptive prop features. Maybe "buttons..." should say "Button controls..." and then "Info..." button would be nested on that page instead?
I don't know. That's where I feel I'd expect it maybe.
Or, instead maybe just swap and edit the buttons so it would be "Prop usage info..." "Button Controls..."
Like If I were making a decision on which to use, I might want to read what features each offers, then check out the buttons after?

Also makes sense to me, I've changed it.

> Anyway, I wish that could be finished up. I have stale submitted PRs waiting for review for a long time now. Almost there.

:fire:
Yeah, that's just your cookies. But indeed, I couldn't leave people with a non-dark-mode-capable Style Editor ;P

>  Running through the end of the tutorial, the config didn't verify successfully:

This is because of an empty define. I have logic (or could've sworn I did) to clear any empty custom options/defines added, but evidently it's not working as intended because you're not the first person I've seen with this issue... will take a look.

> So I clicked OK, which opened the Main Menu window. I chose the config from the dropdown, then the app quit with an error popup "ProffieConfig quit unexpectedly."

Can you explain what happened a bit more? You were in the introduction still when it failed to verify, so you clicked OK and then a new Main Menu opened? The introduction one rose to the top? What happened exactly? When this happened was the introduction window still open and expecting you to do things?

I'll take a look at the crash report (thanks macOS) and see, but some more info from you on what exactly what happening would be good to try and narrow down what happened, as this reads like something that... shouldn't have occurred. (Obviously not the crash, but the stuff around it too)

> Blade Awareness might need some work. I don't think you want to allow both detect and ID to be selected.

Isn't that a valid configuration? Because Blade Detect triggers Blade ID? I've heard Fredrik talk about it on a couple of occasions where he suggests using both for different sabers and installs.

> Enabling just Blade Detect, I can set Blade Detect Pin to something (17), but then can't edit anything else, where I would need to go to the Blade Arrays part and edit a blade_in and blade_out.

This is intended. You select what you want to enable and configure the respective options. For Blade Detect, the only thing to configure is the pin.

It makes sense to me you'd edit the blades on the page you normally edit the blades, just switching the blade array for what you want to edit.

>  blade_out should use NO_BLADE. 

It does. I just can't put a "NO_BLADE" in a number entry. In the future I can look to make this more visually apparent.

>  Enabling Blade ID allows editing of that column, but blade_in doesn't allow editing of the name or ID resistor value (it's stuck at 0)

Yes, this was an oversight (kinda?). I assumed that you'd always want an array for 0 resistance. It would be trivial to make that value editable though... I dunno. To me it makes sense to always have a 0 resistance option, and then more than that could be added by the user. Maybe between the fixed value and non-editable name though that's undesirable?

Iirc half of this was spawned out of the fact that allowing that would have required a lot of extra logic to handle edge cases, and by that point I think I was just burnt out. I might be more willing to take another look at it now, what're your thoughts?

> Yet, adding a blade_out array, both can be edited.

Indeed, this is correct.

>  in the case where the array is to be used when no blade is inserted, NO_BLADE should be the value varable as per how ProffieOS uses it. 

You mean when it's to be used with Blade ID and be used for no blade? (As ofc using NO_BLADE is already done for Blade Detect)

> (changes lately too... there's a range and stuff).

Could you explain more then what this should be? Whenever I set all this up I know it was fairly new, perhaps I've missed changes.

>  ALso, since Enable Power for Blade ID is required, maybe that should default to checkbox ON and greyed out, then allow the pin selection.

It's required? I thought it was only required for neopixels. I could make it default to "ON," but it doesn't make sense to me to force it on. Updating this selection to be like the power pin setup on the blades array page is a todo.

I guess I don't understand here.

**Comment by ryryog25** [comment]
> Still want notes? 

Oh, btw. Yes, always happy to receive feedback :)

**Comment by NoSloppy** [comment]
> > Next, maybe add "seconds" after the PLI, Idle, and Motion timeout values to clarify what the time is in terms of (my brain thought in milliseconds initially)
> 
> These are actually in minutes. I've added labels for them. (Any objections to them being in minutes?)
> 
Yeah, Minutes is good IMO.

> Can you explain what happened a bit more? You were in the introduction still when it failed to verify, so you clicked OK and then a new Main Menu opened? The introduction one rose to the top? What happened exactly? When this happened was the introduction window still open and expecting you to do things?

No, after the failed verify, I dismissed the error popup and finished the introduction.
The "guide" window and the initial main menu window both went away, and a new fresh Main Menu window appeared where the edit stuff window had been.

The crash is due to the empty define being allowed.
I had looked at the custom options during the introduction and was going to add a custom define, but changed my mind. So I guess the close button saved the empty define because the config has a line saying `#define  ` with 2 whitespaces.
That is what's causing the app to crash when it tries to load that saved config from the dropdown.

> 
> > Blade Awareness might need some work. I don't think you want to allow both detect and ID to be selected.
> 
> Isn't that a valid configuration? Because Blade Detect triggers Blade ID? I've heard Fredrik talk about it on a couple of occasions where he suggests using both for different sabers and installs.
> 

Blade Detect is typically just used to toggle between 2 preset arrays based on a latched switch; one for each state of the blade (in or out).
More than one array is not required, and if only one array is used in the config, nothing more really happens other than the bladein.wav and bladeout.wav files get played when the latching event is triggered on and off.

Blade ID can do the same thing as Blade Detect, therefore replacing it with a more robust version, where the resistance is what determines the blade state (no blade inserted should read as infinite resistance and is measured as NO_BLADE (in common/events.h is `#define NO_BLADE 1000000000 /* 1 billion */`)
Here's entries from a config I have for developing the continuous scanning stuff:
```cpp
                //    Use with Blade ID. Specify the LED pad(s) that the main blade uses.
                #define ENABLE_POWER_FOR_ID PowerPINS<bladePowerPin2, bladePowerPin3>

                //    Required when blades on different data pads use common LED power pads.
                //    and/or when BLADE_ID_SCAN_MILLIS is used with Blade ID.
                #define SHARED_POWER_PINS

                //    Choose how often to check for new blade (milliseconds).
                //    Larger values will be a longer delay before a blade is recognized.
                #define BLADE_ID_SCAN_MILLIS 1000
                //    How many Blade ID scans to average (default is 10, 15 works well)
                #define BLADE_ID_TIMES 15

                // This will make the blade ID class for that Measured value range to return as NO_BLADE (use NO_BLADE as the blade definition value.)
                #define NO_BLADE_ID_RANGE 685000,1000000000
```

The range was implemented to help with multi-part hilts where chassis alone, chassis inserted, blade inserted would make up 3 different states, and there was a value for chassis only read as closer to the highest blade value instead of a billion... or something like that. Hard to word, here's the thread if you want to browse:
https://crucible.hubbe.net/t/using-blade-detect-and-bladeid-at-the-same-time/4659


> > Enabling just Blade Detect, I can set Blade Detect Pin to something (17), but then can't edit anything else, where I would need to go to the Blade Arrays part and edit a blade_in and blade_out.
> 
> This is intended. You select what you want to enable and configure the respective options. For Blade Detect, the only thing to configure is the pin.
> 

This is what I was saying about how it should be able to configure 2 blade arrays to be chosen by the blade in or out state. The blade in value can be zero (not reading resistance here), but a blade_out state needs to be set to `NO_BLADE`.
> It makes sense to me you'd edit the blades on the page you normally edit the blades, just switching the blade array for what you want to edit.
> 
> > blade_out should use NO_BLADE.
> 
> It does. I just can't put a "NO_BLADE" in a number entry. In the future I can look to make this more visually apparent.
> 
> > Enabling Blade ID allows editing of that column, but blade_in doesn't allow editing of the name or ID resistor value (it's stuck at 0)
> 
> Yes, this was an oversight (kinda?). I assumed that you'd always want an array for 0 resistance. 
There would never be zero resistance.
> 
> > ALso, since Enable Power for Blade ID is required, maybe that should default to checkbox ON and greyed out, then allow the pin selection.
> 
> It's required? I thought it was only required for neopixels. I could make it default to "ON," but it doesn't make sense to me to force it on. Updating this selection to be like the power pin setup on the blades array page is a todo.
> 
> I guess I don't understand here.
I've never used a Proffieboard with an inhilt LED blade, but regardless of what type of setup, the reading of resistance can only occur if the FETs are on. Therefore, Blade ID checked on should likely default the Enable Power to ON and uneditable.


Next observation:
 - I assume having Arduino open is competition for the port and that's why the board isn't available to the app? It hangs on "Fetching Devices..." . The close button doesn't cancel it, nor does switching to Arduino and quitting help.  Need to force quit ProffieConfig.
 
 - I love the darkmode Serial Monitor. It's going to be my new goto for sure. Any chance there could be an option to change the line spacing? I have a ton of messages, that's a lot of scrolling.
 
 - I only glanced at the help menu, which opens markdown git pages which is great. But the visual, hands-on walkthrough of the introduction is helpful in a different way. Do you think you could make it so that a re-run (not the first run) could allow jumping forward and backward though the tutorial? I'm imagining a sidebar with each step as a link that could be jumped to at any point. Having to go through each step and creating new configs for no reason just to get to the "blade awareness" section for example is a bit tedious.

**Comment by ryryog25** [comment]
> No, after the failed verify, I dismissed the error popup and finished the introduction.
The "guide" window and the initial main menu window both went away, and a new fresh Main Menu window appeared where the edit stuff window had been.
> 
> The crash is due to the empty define being allowed.
I had looked at the custom options during the introduction and was going to add a custom define, but changed my mind. So I guess the close button saved the empty define because the config has a line saying #define  with 2 whitespaces.
That is what's causing the app to crash when it tries to load that saved config from the dropdown.

Ok, gotcha, yeah, I took a look at the crash report. Indeed, the crash occurs when it tries to read the config again. It seems that I made assumptions about define stuff being there, so I'll need to correct that (as a broken config could also crash it this way), and also fix the fact that it's saving empty defines.

> Blade Detect is typically just used to toggle between 2 preset arrays based on a latched switch; one for each state of the blade (in or out).
More than one array is not required, and if only one array is used in the config, nothing more really happens other than the bladein.wav and bladeout.wav files get played when the latching event is triggered on and off.

I didn't know more than one array wasn't required, noted. (Further thoughts on that later in this message)

> Blade ID can do the same thing as Blade Detect, therefore replacing it with a more robust version, where the resistance is what determines the blade state (no blade inserted should read as infinite resistance and is measured as NO_BLADE (in common/events.h is #define NO_BLADE 1000000000 /* 1 billion */)

Right, that much I understood. So is the goal to largely move away from Blade Detect? I'm still confused on why I'd want to outright disallow using both... I don't think there's conflict between them.

Perhaps I'm misunderstanding intent and things are going to be reworked in the future to be a particular way?

> The range was implemented to help with multi-part hilts where chassis alone, chassis inserted, blade inserted would make up 3 different states, and there was a value for chassis only read as closer to the highest blade value instead of a billion... or something like that. Hard to word, here's the thread if you want to browse:

Ah, yeah, I watched that thread when it was "developing," but I guess I missed the outcome of that range define. So I'll need to add that, thanks!

> This is what I was saying about how it should be able to configure 2 blade arrays to be chosen by the blade in or out state. The blade in value can be zero (not reading resistance here), but a blade_out state needs to be set to NO_BLADE.

:+1: So then stuff in that regard is set up correctly in ProffieConfig, however the fact that a `0` is displayed instead is evidently confusing (I'm not sure why I thought it wouldn't be... I was probably just burnt out and lazy to be completely honest, so that's something I'll rectify)

> There would never be zero resistance.

I understand, but as I understand Blade ID, it chooses the nearest resistance in your blade arrays. If your only two blade arrays are 0 resistance and NO_BLADE, when you have anything inserted, that'd bring it closer to 0 than NO_BLADE, (unless your range was set very low), right?

Ultimately it seems like the problems w/ Blade ID and Blade Detect are as follows (please lmk if I missed something):
1. Inability to customize `blade_in` and `blade_out` default blade arrays.
1. Inability to simply only have one array/ more control over the arrays for the user.
1. Not being able to choose `NO_BLADE` as a value, and its use not being visible for `blade_out`
1. Lack of support for `NO_BLADE_ID_RANGE`
1. Enable Power for ID is not defaulted to On. 

Given the current work on ProffieConfig V2, and because I know how messy the configuration reading and handling has gotten in V1, these are things I might just wait to implement properly in the rewrite, but I'll definitely keep them in mind to do a proper job of setting them up for the next version...

> I've never used a Proffieboard with an inhilt LED blade, but regardless of what type of setup, the reading of resistance can only occur if the FETs are on. Therefore, Blade ID checked on should likely default the Enable Power to ON and uneditable.

I guess I was imagining if you were to use Blade ID for chassis connectors. Then Blade ID could be resistored in the upper half of a saber to a proper GND, and you wouldn't need to enable the MOSFETs to do that scanning. I can default the Enable Power for ID to on, but I still feel like it's a bad idea to make it not an option that can be turned off.

> I assume having Arduino open is competition for the port and that's why the board isn't available to the app? It hangs on "Fetching Devices..." . The close button doesn't cancel it, nor does switching to Arduino and quitting help. Need to force quit ProffieConfig.

On Windows it's a known issue that having Arduino open causes issues with the upload, however refreshing still works fine, and it never hangs, so this seems to either be some macOS specific goodness and/or there's a different bug going on. I'll try and take a look...

> The close button doesn't cancel it,

Yeah, this is (kind of) a wxWidgets thing, it always shows the close button even if I don't want it... at present I don't have a great way to close things and force everything to stop cleanly (so that button doesn't do anything, even normally), but perhaps that's something to look into, bug/hang aside.

> I love the darkmode Serial Monitor. It's going to be my new goto for sure. Any chance there could be an option to change the line spacing? I have a ton of messages, that's a lot of scrolling.

The spacing and formatting in general is an odd thing in the macOS and Linux versions. The serial monitor is actually written from scratch for these platforms (because I could and I wanted to try), whereas on Windows (because I couldn't figure it out when I was first setting it up), it just opens the arduino-cli serial monitor.

I've learned a fair bit about how the port needs to be configured since then, and I could probably fix that up. (mostly because I had to handle uploading manually without Arduino on the Windows version, so I got to write all that and figure it out)

>  I only glanced at the help menu, which opens markdown git pages which is great. But the visual, hands-on walkthrough of the introduction is helpful in a different way. Do you think you could make it so that a re-run (not the first run) could allow jumping forward and backward though the tutorial? I'm imagining a sidebar with each step as a link that could be jumped to at any point. Having to go through each step and creating new configs for no reason just to get to the "blade awareness" section for example is a bit tedious.

Perhaps... ;)

 What sort of things would you expect the introduction to cover then? For the blade awareness in particular, I don't hardly give much info on it during the introduction (Another case of being burnt out after a very long subproject), so just going back to that text during the introduction I don't feel would be super useful.
 
Or are you saying just in general simply the ability to see how the introduction takes you through the setup process, but starting from a particular point?

**Comment by NoSloppy** [comment]
> > No, after the failed verify, I dismissed the error popup and finished the introduction.
> > The "guide" window and the initial main menu window both went away, and a new fresh Main Menu window appeared where the edit stuff window had been.
> > The crash is due to the empty define being allowed.
> > I had looked at the custom options during the introduction and was going to add a custom define, but changed my mind. So I guess the close button saved the empty define because the config has a line saying #define  with 2 whitespaces.
> > That is what's causing the app to crash when it tries to load that saved config from the dropdown.
> 
> Ok, gotcha, yeah, I took a look at the crash report. Indeed, the crash occurs when it tries to read the config again. It seems that I made assumptions about define stuff being there, so I'll need to correct that (as a broken config could also crash it this way), and also fix the fact that it's saving empty defines.
> 
> > Blade Detect is typically just used to toggle between 2 preset arrays based on a latched switch; one for each state of the blade (in or out).
> > More than one array is not required, and if only one array is used in the config, nothing more really happens other than the bladein.wav and bladeout.wav files get played when the latching event is triggered on and off.
> 
> I didn't know more than one array wasn't required, noted. (Further thoughts on that later in this message)
> 
> > Blade ID can do the same thing as Blade Detect, therefore replacing it with a more robust version, where the resistance is what determines the blade state (no blade inserted should read as infinite resistance and is measured as NO_BLADE (in common/events.h is #define NO_BLADE 1000000000 /* 1 billion */)
> 
> Right, that much I understood. So is the goal to largely move away from Blade Detect? I'm still confused on why I'd want to outright disallow using both... I don't think there's conflict between them.
> 
> Perhaps I'm misunderstanding intent and things are going to be reworked in the future to be a particular way?
> 

I just don't know why you'd need both, although I guess blade ID could be used for main blade while blade detect used for crystal present or not, or chassis insertion or anything else. I just know that since we developed the continuous scanning, I no longer need to use blade detect since ID covers the same, most basic use, with the benefit of being able to use on an already-built hilt that was not wired with a dedicated pin for latching detection.
I'd kinda have to get my head back into that function to remember more details or have a really clear answer.
Of course, Fredrik likely could answer anything if you want to ask him. I just have limited active buffer space in my brain. Lately it's filled with dual blade stuff. Before that was sound font converter site, before that OLED stuff, before that Style Editor... each project's details fade a bit as I walk away from it.

> > There would never be zero resistance.
> 
> I understand, but as I understand Blade ID, it chooses the nearest resistance in your blade arrays. If your only two blade arrays are 0 resistance and NO_BLADE, when you have anything inserted, that'd bring it closer to 0 than NO_BLADE, (unless your range was set very low), right?

sounds right.


> 
> Ultimately it seems like the problems w/ Blade ID and Blade Detect are as follows (please lmk if I missed something):
> 
> 1. Inability to customize `blade_in` and `blade_out` default blade arrays.
> 2. Inability to simply only have one array/ more control over the arrays for the user.
> 3. Not being able to choose `NO_BLADE` as a value, and its use not being visible for `blade_out`
> 4. Lack of support for `NO_BLADE_ID_RANGE`
> 5. Enable Power for ID is not defaulted to On.
> 

I'd have to run through it again with my blade detection hat on (see above)

> Given the current work on ProffieConfig V2, 
Curious what is in store for v2. I saw you posted on Discord a screenshot of code. It was transitions. Are you making a built-in style editor/emulator?


> Or are you saying just in general simply the ability to see how the introduction takes you through the setup process, but starting from a particular point?

Yeah, like to get to subblade part, user needs to go through all the previous steps first. Being able to jump right to that step could be useful once the initial hand-holding is complete.

**Comment by ryryog25** [comment]
> I'd kinda have to get my head back into that function to remember more details or have a really clear answer.
Of course, Fredrik likely could answer anything if you want to ask him.

Haha, gotcha. Yeah, asking Fredrik probably wouldn't be a bad idea at some point. I'll set an internal reminder to bug him about it.

> Yeah, like to get to subblade part, user needs to go through all the previous steps first. Being able to jump right to that step could be useful once the initial hand-holding is complete.

Gotcha, noted. :)

> Curious what is in store for v2. I saw you posted on Discord a screenshot of code. It was transitions. Are you making a built-in style editor/emulator?

Indeed, among other things. Originally my plan was to simply use a webview to embed the style preview and I actually setup some JS injection to allow me to format everything and set things so that it worked quite well... then I tested on macOS, and it worked well... then I tested on Windows... and because of the way Windows uses MS Edge for webviews, I spent weeks trying to get it to work and just never could.

Ultimately, I decided that since I already had the framework for all the styles laid out, I might as well add the actual code to them too and, in the spirit of the project as a proper desktop application, figure out how to setup rendering with OpenGL.

You can see some of the development as it happens on the `dev` branch if you're curious and have the time :)

Additionally, style management in general will be a lot more refined, the style editor I created already has the ability to seamlessly import and export the styles, so I'm going to internally manage styles in ProffieConfig, allow for selecting the same styles cross-config, and handle including the necessary ones.

The introduction in V2 will be much shorter, because adding a new config will (with the option to skip) have a guided setup that my plan is to go over wiring (since right now only Fredrik's generator does wiring), and setup from a physical standpoint first, since that also seems to trouble people, and logically progress from there.

Also the ability to import a config just for it's wiring configuration, and edit it in that initial setup too.

There's also plans for SD card and soundfont management, making sure all that aligns. Smoother handling of the `common` folder, etc.

Configurations are going to be stored in `pconf` in V2 to allow for special annotation and metadata (don't worry, they'll still be exportable... otherwise I couldn't use them for uploads anyways!), and there's some tricks and ideas I have for features that will come with that.

There's also other things planned as I move to creating V2's system for uploads... and, well... I won't go over everything planned out, but suffice to say, that's not everything, just a few highlights, and I'm super excited!

**Comment by ryryog25** [comment]
The ProffieConfig Style Editor, since you asked about it specifically, will be scratch-esque to make them easy for newcomers. It makes little sense for me to recreate the ProffieOS Style Editor, that's not my goal. And because of the fact that imports and exports of styles are as easy as copy/paste (into and out of block form), there's nothing stopping anyone from doing things in the ProffieOS style editor if they prefer (or getting them from any other source... Fett's website for example) and then pasting them into ProffieConfig in order to use those, manage them, and edit them.

The actual style editor itself as far as its core functions is complete and works quite well. Polish things and filling out are obviously still in progress, but proof of concept has been achieved, so it's a guaranteed feature.

There will also be built in documentation, with every block having a "?" button that jumps directly to the documentation for more info. Everything is color-coded, there's search functionality for all the styles, and overall the goal is to make getting into creating your own styles very easy and straightforward.

It's no wonder there's so few people doing it themselves with the stuff as it stands, so my goal is to try and get people over that hurdle and just make it easy.

That's the mission of ProffieConfig as a whole after all, make it all easy so the ideas and concepts of ProffieOS are not lost should someone want to dig deeper (and in fact, ProffieConfig in a perfect form should allow all possible customizations of ProffieOS in an easy-to-use way), but it still encourages and welcomes people who are complete novices to the world of proffie, arduino, or otherwise.

**Comment by ryryog25** [comment]
With a near 100% chance of having numerous bugs I'll get to fix once I get everything else working to test, got all the transitions mapped: https://github.com/ryryog25/ProffieConfig/blob/dev/src/styles/elements/transitions.cpp
